components/layout_2020/flexbox/layout.rs

```diff
@@ -2,32 +2,1090 @@
+            box_,
+            tree_rank,
+            content_box_size,
+            content_min_size,
+            content_max_size,
+            padding,
+            border,
+            margin,
+            pbm_auto_is_zero,
+            flex_base_size,
+            hypothetical_main_size,
+        }
+    }
+}
+
+/// https://drafts.csswg.org/css-flexbox/#min-size-auto
+fn automatic_min_size(_box: &IndependentFormattingContext) -> Length {
+    // FIMXE: implement the actual algorithm
+    Length::zero() // Give an incorrect value rather than panicking
+}
+
+/// https://drafts.csswg.org/css-flexbox/#algo-main-item
+fn flex_base_size(
+    flex_context: &FlexContext,
+    flex_item: &mut IndependentFormattingContext,
+    cross_axis_is_item_block_axis: bool,
+    content_box_size: FlexRelativeVec2<LengthOrAuto>,
+    padding_border_sums: FlexRelativeVec2<Length>,
+) -> Length {
+    let used_flex_basis = match &flex_item.style().get_position().flex_basis {
+        FlexBasis::Content => FlexBasis::Content,
+        FlexBasis::Size(Size::LengthPercentage(length_percentage)) => {
+            let apply_box_sizing = |length: Length| {
+                match flex_item.style().get_position().box_sizing {
+                    BoxSizing::ContentBox => length,
+                    BoxSizing::BorderBox => {
+                        // This may make `length` negative,
+                        // but it will be clamped in the hypothetical main size
+                        length - padding_border_sums.main
+                    },
+                }
+            };
+            // “For example, percentage values of flex-basis are resolved
+            //  against the flex item’s containing block (i.e. its flex container);”
+            match flex_context.container_definite_inner_size.main {
+                Some(container_definite_main_size) => {
+                    let length = length_percentage
+                        .0
+                        .percentage_relative_to(container_definite_main_size);
+                    FlexBasis::Size(apply_box_sizing(length))
+                },
+                None => {
+                    if let Some(length) = length_percentage.0.to_length() {
+                        FlexBasis::Size(apply_box_sizing(length))
+                    } else {
+                        // “and if that containing block’s size is indefinite,
+                        //  the used value for `flex-basis` is `content`.”
+                        // https://drafts.csswg.org/css-flexbox/#flex-basis-property
+                        FlexBasis::Content
+                    }
+                },
+            }
+        },
+        FlexBasis::Size(Size::Auto) => {
+            // “When specified on a flex item, the `auto` keyword retrieves
+            //  the value of the main size property as the used `flex-basis`.”
+            match content_box_size.main {
+                LengthOrAuto::LengthPercentage(length) => FlexBasis::Size(length),
```

Comment:
should percentages be resolved relative to something here?

Reply:
`content_box_size` here originally comes from the `ComputedValuesExt` method of the same name, which already takes care of resolving percentages (but not `auto`)